### PR TITLE
[ExaquteSandboxApplication] Add divergence computation process and python process

### PR DESCRIPTION
**Pull request by riccardotosi**
This PR adds the divergence process computation, needed for calling the [time averaging process](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/processes/time_averaging_process.cpp). This process takes divergence computations from [this process](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ExaquteSandboxApplication/custom_processes/weighted_divergence_calculation_process.cpp), which will be removed in a successive PR.

The python process for performing adaptive remeshing using a divergence free metric is added, updating #5147 (which can be now closed).

For both the divergence process and the adaptive remeshing python process a test is added.

### CODE REVIEW

**Comment by riccardotosi** [comment]
Ping

File: applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py
Comment (marcnunezc) [inline applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py]: I think that it's better to do this check in the specific test, using the `self.skip_test` method

For instance:
https://github.com/KratosMultiphysics/Kratos/blob/24a655a28a0c8aca15c6d85f7da1e2986affa5d8/applications/CompressiblePotentialFlowApplication/tests/potential_flow_test_factory.py#L43-L45

With your current code, the test suite wouldn't recognize it as an skipped test I think.

File: applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py
Comment (marcnunezc) [inline applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py]: Do you need these lines? Are these test the same that have been added in the small suite?

File: applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py]: Done

File: applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py]: If I do not add this, the test `TimeAveragingProcessTests` does not run

Comment (marcnunezc) [reply applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py]: Try adding these lines instead, it should work I think:
```

    allSuite = suites['all']
    allSuite.addTests(nightSuite)
```

Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/tests/test_ExaquteSandboxApplication.py]: Done, it works

File: applications/ExaquteSandboxApplication/custom_python/exaqute_sandbox_python_application.cpp
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_python/exaqute_sandbox_python_application.cpp:47]: There's already a `DIVERGENCE` variable in the FluidDynamicsApplication. I'm wondering if it would make sense to migrate it to the core.

File: applications/ExaquteSandboxApplication/tests/test_divergencefree_refinement_process.py
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/tests/test_divergencefree_refinement_process.py]: You can directly construct the process you are testing to avoid all these `self.process_list` loops.

File: applications/ExaquteSandboxApplication/tests/cpp_tests/test_divergence_process.cpp
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/tests/cpp_tests/test_divergence_process.cpp]: Do you really need this? The element you are creating is the prototype one, which does nothing.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: For the sake of consistency
```suggestion
    class KRATOS_API(EXAQUTE_SANDBOX_APPLICATION) CalculateDivergenceProcess : public Process
```

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: I suggest to create another constructor taking as arguments model + parameters. This would enable it's usage from the GUI.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Doxygen documentation is missing.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Same.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (rubenzorrilla) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: I also think that these two methods can be `private`.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]:
```suggestion
        DivergenceProcess(DivergenceProcess const& rOther) = delete;
```

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: I think you can include this one in the cpp.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: I think you can include this one in the cpp.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: I think you can include this one in the cpp.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: At first glance, I think this can be `const`.
```suggestion
            const auto& r_geometry = it_elem->GetGeometry();
```

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: You are checking the if statement for each node. You can check it once per element by putting it outside the nodes loop and creating an auxiliary method for the 2D and 3D cases.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: You are allocating `grad_x`,`grad_y` and `grad_z` lots of times. You can  make it more efficient by doing this outside the main loop and passing them inside the `firstpravate` clause.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: I think you can skip this if statement by passing an `std::vector<Vector>` and doing the corresponding summation in a loop.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: Same comment as above.

File: applications/ExaquteSandboxApplication/custom_python/add_custom_processes_to_python.cpp
Comment (rubenzorrilla) [inline applications/ExaquteSandboxApplication/custom_python/add_custom_processes_to_python.cpp:53]: Take into account the aforementioned naming convention issue.

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
Some performance, convention and documentation minor issues.

File: applications/ExaquteSandboxApplication/tests/test_divergencefree_refinement_process.py
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/tests/test_divergencefree_refinement_process.py]: Done.

File: applications/ExaquteSandboxApplication/tests/cpp_tests/test_divergence_process.cpp
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/tests/cpp_tests/test_divergence_process.cpp]: Removed.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Done the renaming.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Added and moved to private.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Done.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Done.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Done.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Done.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Done.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: Done.

File: applications/ExaquteSandboxApplication/custom_python/add_custom_processes_to_python.cpp
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_python/add_custom_processes_to_python.cpp:53]: Done.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: I can actually do this operation independently from the dimension and remove the if statement.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: I removed the if statement by knowing the grad_z is zero for 2-dimensional problems.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: Also here I removed the if statement by knowing the grad_z is zero for 2-dimensional problems.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: Done, thanks for the hint!

**Comment by riccardotosi** [comment]
@rubenzorrilla thanks for the comments. I have applied the requested changes, apart from the first one regarding the `DIVERGENCE` variable.

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.h
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.h]: Done.

File: applications/ExaquteSandboxApplication/custom_python/exaqute_sandbox_python_application.cpp
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_python/exaqute_sandbox_python_application.cpp:47]: I think it is a good idea, I would propose to open a PR to do this after merging this PR.

**Comment by riccardotosi** [comment]
@rubenzorrilla ping

**Review by rubenzorrilla — DISMISSED** [review]

**Comment by rubenzorrilla** [comment]
Take Codacy suggestions into account.

**Comment by riccardotosi** [comment]
Done @rubenzorrilla

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (marcnunezc) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: Following @rubenzorrilla suggestions and for the sake of consistency I would also apply the convention to the filename:
divergence_process.cpp -> calculate_divergence_process.cpp
divergence_process.h -> calculate_divergence_process.h

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (marcnunezc) [inline applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: Also from @rubenzorrilla suggestions, we can also put the declaration of these variables outside the for loop.

**Review by marcnunezc — COMMENTED** [review]
Apart from the comments below I'm OK with everything. We need approval from @rubenzorrilla anyway!

File: applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp
Comment (riccardotosi) [reply applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp]: Correct, I forgot to change.

**Review by marcnunezc — APPROVED** [review]
Ok thanks!

**Pull request merged**
